velodrome/token-counter/token-counter.go

```diff
@@ -23,24 +23,72 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
+	"net/http"
 	"os"
 	"path/filepath"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/golang/glog"
 	"github.com/google/go-github/github"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 	"golang.org/x/oauth2"
 )
 
 type tokenCounterFlags struct {
-	influx InfluxConfig
-	tokens []string
+	influx      InfluxConfig
+	tokens      []string
+	metricsPort int
+	host        string
+	promOnly    bool
+}
+
+func NewCollector(tokens []TokenHandler) prometheus.Collector {
+	return &Collector{
+		tokens: tokens,
+		tokenCounter: prometheus.NewDesc(
+			"github_token_count",
+			"Number of counted API calls against the github API within a reset window",
+			[]string{"login", "rate"}, nil),
+	}
+}
+
+type Collector struct {
+	tokenCounter *prometheus.Desc
+	tokens       []TokenHandler
+}
+
+func (c *Collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.tokenCounter
+}
+
+func (c *Collector) Collect(ch chan<- prometheus.Metric) {
+	var wg sync.WaitGroup
+	wg.Add(len(c.tokens))
+	for _, handler := range c.tokens {
+		go func(handler TokenHandler) {
+			defer wg.Done()
+			rate, err := handler.ProcessNow()
+			if err != nil {
+				logrus.WithError(err).WithField("login", handler.login).Errorf("Failed to fetch rate limits for login '%s'", handler.login)
+			} else {
+				ch <- prometheus.MustNewConstMetric(c.tokenCounter, prometheus.GaugeValue, float64(rate.Limit-rate.Remaining), handler.login, "used")
+				ch <- prometheus.MustNewConstMetric(c.tokenCounter, prometheus.GaugeValue, float64(rate.Remaining), handler.login, "remaining")
```

Comment:
Is `Collect` only called once when the process starts or is it called periodically? `Must` variants of functions should be avoided at any time besides initialization (and maybe always) because they panic on failure.